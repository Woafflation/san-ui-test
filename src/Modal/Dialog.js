import React from 'react'
import cx from 'classnames'
import Modal from './Modal'
import Panel from '../Panel/Panel'
import Icon from '../Icon'
import Button from '../Button'
import styles from './Modal.module.scss'

const Dialog = ({ classes, title, children, showCloseBtn, ...props }) => (
  <Modal
    {...props}
    classes={{
      wrapper: styles.dialogModal,
      modal: cx(styles.dialog, classes.dialog)
    }}
    as={Panel}
  >
    {closeModal => (
      <>
        <Panel.Title
          withPadding
          className={cx(styles.dialog__title, classes.title)}
        >
          {title}
          {showCloseBtn && (
            <Icon
              type='close'
              onClick={closeModal}
              className={styles.dialog__close}
            />
          )}
        </Panel.Title>
        {children}
      </>
    )}
  </Modal>
)

Dialog.ScrollContent = ({ className, ...props }) => (
  <Panel.Content className={cx(styles.dialog__content, className)} {...props} />
)

Dialog.Actions = ({ className, ...props }) => (
  <div className={cx(styles.dialog__actions, className)} {...props} />
)

Dialog.Cancel = props => <Button border {...props} />

Dialog.Approve = props => <Button variant='fill' accent='positive' {...props} />

Dialog.defaultProps = {
  showCloseBtn: true,
  classes: {
    dialog: '',
    title: ''
  }
}

export default Dialog
