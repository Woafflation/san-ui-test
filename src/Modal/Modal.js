import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import cx from 'classnames'
import styles from './Modal.module.scss'
/**
 * "Modal" is a primitive for building higher level abstractions like "Dialog".
 */
class Modal extends Component {
  state = {
    open: this.props.defaultOpen || this.props.open,
    showCloseAnimation: false,
    // to prevent false-click and close dialog. Useful for popup, that appears on non-user clicks (after timeout) or several fast clicks
    closeOnDimmed: !this.props.preventCloseOnDimmedFromStart
  }

  componentDidMount () {
    window.addEventListener('keyup', this.onKeyUp)
  }

  componentDidUpdate ({ open: prevOpen }, prevState) {
    if (
      this.props.preventCloseOnDimmedFromStart &&
      prevState.open !== this.state.open &&
      this.state.open
    ) {
      this.closeTimer = setTimeout(
        () => this.setState({ closeOnDimmed: true }),
        1000
      )
    }

    const isControlled = typeof this.props.open !== 'undefined'

    if (isControlled && this.props.open !== prevOpen) {
      if (this.props.open) {
        this.setState({ open: true })
      } else if (this.props.withAnimation) {
        if (!this.state.showCloseAnimation) {
          this.setState({ showCloseAnimation: true })
          setTimeout(
            () => this.setState({ open: false, showCloseAnimation: false }),
            150
          )
        }
      } else {
        this.setState({ open: false })
      }
    }

    if (
      !this.state.open &&
      this.props.preventCloseOnDimmedFromStart &&
      this.state.closeOnDimmed
    ) {
      this.setState({ closeOnDimmed: false })
    }
  }

  componentWillUnmount () {
    window.removeEventListener('keyup', this.onKeyUp)
    clearTimeout(this.closeTimer)
  }

  openModal = () => {
    this.setState({ open: true }, this.props.onOpen)
  }

  closeModal = () => {
    const isControlled = typeof this.props.open !== 'undefined'
    if (isControlled) {
      this.props.onClose()
    } else {
      if (this.props.withAnimation) {
        this.setState({ showCloseAnimation: true })
        setTimeout(
          () =>
            this.setState(
              { open: false, showCloseAnimation: false },
              this.props.onClose
            ),
          150
        )
      } else {
        this.setState({ open: false }, this.props.onClose)
      }
    }
  }

  onDimmedClose = () => {
    if (this.state.closeOnDimmed) {
      this.closeModal()
    }
  }

  onKeyUp = ({ code }) => {
    if (code === 'Escape' && this.state.open) {
      this.closeModal()
    }
  }

  render () {
    const { open, showCloseAnimation } = this.state
    const {
      trigger,
      children,
      classes,
      as: El,
      withAnimation,
      passOpenStateAs,
      modalProps = {}
    } = this.props

    const render =
      typeof children === 'function' ? children(this.closeModal) : children

    return (
      <>
        {trigger &&
          React.cloneElement(trigger, {
            onClick: trigger.props.onClick || this.openModal,
            [passOpenStateAs]: passOpenStateAs ? open : undefined
          })}
        {open &&
          ReactDOM.createPortal(
            <div className={cx(styles.wrapper, classes.wrapper)}>
              <El
                className={cx(
                  styles.modal,
                  withAnimation && styles.modal__withAnimation,
                  withAnimation && showCloseAnimation && styles.hide,
                  classes.modal
                )}
                {...modalProps}
              >
                {render}
              </El>
              <div
                className={cx(
                  styles.dimmed,
                  withAnimation && styles.dimmed__withAnimation,
                  classes.bg
                )}
                onClick={this.onDimmedClose}
              />
            </div>,
            document.body
          )}
      </>
    )
  }
}

Modal.defaultProps = {
  onClose: () => {},
  onOpen: () => {},
  as: 'div',
  classes: { wrapper: '', modal: '', bg: '' },
  open: undefined,
  defaultOpen: false
}

Modal.propTypes = {
  trigger: PropTypes.node,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,

  /** Used for controlling modal from outside*/
  open: PropTypes.bool,
  defaultOpen: PropTypes.bool
}

export default Modal
