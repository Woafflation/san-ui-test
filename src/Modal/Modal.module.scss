@import '../variables.scss';

.wrapper {
  position: fixed;
  z-index: 1000;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  overflow-y: auto;
  overflow-x: hidden;
}

.dimmed {
  position: fixed;
  z-index: -1;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(0, 0, 0, 0.8);
}

.dialogModal {
  display: flex;
  align-items: center;
  justify-content: center;
}

.close {
  position: absolute;
  cursor: pointer;
  color: var(--waterloo);
  left: -16px;
  top: -16px;
}

.dialog {
  max-width: 85%;
  max-height: 85%;
  display: flex;
  flex-direction: column;

  &__actions {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    padding: 20px;
  }

  &__close {
    cursor: pointer;
    fill: var(--waterloo);
  }

  &__title {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  &__container {
    max-width: 85%;
    margin: 5% auto;
  }

  &__content {
    overflow-y: auto;
    overflow-x: hidden;
  }
}
