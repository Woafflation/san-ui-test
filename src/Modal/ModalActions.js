import React from "react";
import cx from 'classnames';
import PropTypes from "prop-types";
import styles from "./Modal.module.scss";

const ModalActions = ({ className, children })  => (
  <div className={cx(styles.actions, className)}>
    {children}
  </div>
)

ModalActions.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
};

export default ModalActions
