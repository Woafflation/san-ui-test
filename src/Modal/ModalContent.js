import React from "react";
import cx from 'classnames';
import PropTypes from "prop-types";
import styles from "./Modal.module.scss";

const ModalContent = ({ className, children })  => (
  <div className={cx(styles.content, className)}>
    {children}
  </div>
)

ModalContent.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired
}

export default ModalContent
