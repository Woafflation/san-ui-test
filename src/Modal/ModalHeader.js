import React from 'react'
import PropTypes from 'prop-types'
import Icon from '../Icon'
import styles from './Modal.module.scss'

const ModalHeader = ({ onCloseModal, hideCloseIcon, title }) => (
  <div className={styles.header}>
    <div className={styles.title}>{title}</div>
    {hideCloseIcon || (
      <Icon onClick={onCloseModal} className={styles.closeIcon} type='close' />
    )}
  </div>
)

ModalHeader.defaultProps = {
  hideCloseIcon: false
}

ModalHeader.propTypes = {
  title: PropTypes.string.isRequired,
  hideCloseIcon: PropTypes.bool,
  onCloseModal: PropTypes.func
}

export default ModalHeader
