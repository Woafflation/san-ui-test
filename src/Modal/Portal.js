import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import styles from "./Modal.module.scss";

class Portal extends React.Component {
  constructor() {
    super();

    this.portal = React.createRef();
  }

  componentDidMount() {
    this.portal.current.focus();
  }

  handleOnKeyUp = e => {
    if (e.keyCode == 27) {
      this.props.closeModal();
    }
  };

  handleOnClick = e => {
    if (this.portal.current == e.target) {
      this.props.closeModal();
    }
  };

  render() {
    const { children } = this.props;

    return ReactDOM.createPortal(
      <div
        tabIndex={0}
        ref={this.portal}
        className={styles.container}
        onClick={this.handleOnClick}
        onKeyUp={this.handleOnKeyUp}
      >
        <div className={styles.dialog}>
          {children}
        </div>
      </div>,
      document.body
    );
  }
}

Portal.propTypes = {
  children: PropTypes.node.isRequired,
  closeModal: PropTypes.func.isRequired,
};

export default Portal;
