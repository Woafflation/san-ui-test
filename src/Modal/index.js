export { default } from './Modal'
export { default as ModalContent } from './ModalContent'
export { default as ModalHeader } from './ModalHeader'
export { default as ModalActions } from './ModalActions'
