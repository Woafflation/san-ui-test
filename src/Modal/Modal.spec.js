/* eslint-env jest */
import React from 'react'
import toJson from 'enzyme-to-json'
import { shallow, mount } from 'enzyme'
import Button from './../Button'
import Modal from './Modal'

describe('Modal component', () => {
  let mockCb

  beforeEach(() => {
    mockCb = jest.fn()
  })

  it('smoke', () => {
    const wrapper = mount(
      <Modal
        trigger={<Button>Check</Button>}
        title='Any title'
        onConfirmClick={mockCb}
      >
        Any internal modal text
      </Modal>
    )
    wrapper.find('button').simulate('click')
    const actions = wrapper.find('ModalActions')
    expect(actions.exists()).toBe(true)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should have right title', () => {
    const wrapper = mount(
      <Modal
        trigger={<Button>Check</Button>}
        title='Any title'
        showDefaultActions={false}
        onConfirmClick={mockCb}
      >
        Any internal modal text
      </Modal>
    )

    wrapper.find('button').simulate('click')
    const title = wrapper.find('.title').text()
    expect(title).toBe('Any title')
  })

  it("should don't have default actions, if pass showDefaultActions = false", () => {
    const wrapper = mount(
      <Modal
        trigger={<Button>Check</Button>}
        title='Any title'
        showDefaultActions={false}
      >
        Any internal modal text
      </Modal>
    )

    wrapper.find('button').simulate('click')
    const actions = wrapper.find('ModalActions')
    expect(actions.exists()).toBe(false)
  })
})
