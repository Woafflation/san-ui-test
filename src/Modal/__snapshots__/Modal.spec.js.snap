// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component smoke 1`] = `
<Modal
  hideCloseIcon={false}
  onClose={[Function]}
  onConfirmClick={[MockFunction]}
  onOpen={[Function]}
  showDefaultActions={true}
  title="Any title"
  trigger={
    <Button
      as="button"
      border={false}
      className=""
      fluid={false}
      isActive={false}
    >
      Check
    </Button>
  }
>
  <Button
    as="button"
    border={false}
    className=""
    fluid={false}
    isActive={false}
    onClick={[Function]}
  >
    <button
      className=" button undefined"
      onClick={[Function]}
    >
      Check
    </button>
  </Button>
  <Portal
    containerInfo={
      <div
        id="ui-modal"
      >
        <div
          class="dialog panel modal"
        >
          <div
            class="header"
          >
            <div
              class="title"
            >
              Any title
            </div>
            <svg
              class="closeIcon"
              height="12"
              viewBox="0 0 12 12"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.7803 1.28033C12.0732 0.987437 12.0732 0.512563 11.7803 0.21967C11.4874 -0.0732233 11.0126 -0.0732233 10.7197 0.21967L6 4.93934L1.28033 0.21967C0.987437 -0.0732233 0.512563 -0.0732233 0.21967 0.21967C-0.0732233 0.512563 -0.0732233 0.987437 0.21967 1.28033L4.93934 6L0.21967 10.7197C-0.0732233 11.0126 -0.0732233 11.4874 0.21967 11.7803C0.512563 12.0732 0.987437 12.0732 1.28033 11.7803L6 7.06066L10.7197 11.7803C11.0126 12.0732 11.4874 12.0732 11.7803 11.7803C12.0732 11.4874 12.0732 11.0126 11.7803 10.7197L7.06066 6L11.7803 1.28033Z"
              />
            </svg>
          </div>
          <div
            class="content"
          >
            Any internal modal text
            <div
              class="actions"
            >
              <button
                class="btnCancel button undefined negative bordered"
              >
                Cancel
              </button>
              <button
                class=" button fill positive"
              >
                Confirm
              </button>
            </div>
          </div>
        </div>
        <div
          class="dimmed"
        />
      </div>
    }
  >
    <Panel
      className="dialog"
      padding={false}
      variant="modal"
    >
      <div
        className="dialog panel modal"
      >
        <ModalHeader
          hideCloseIcon={false}
          onCloseModal={[Function]}
          title="Any title"
        >
          <div
            className="header"
          >
            <div
              className="title"
            >
              Any title
            </div>
            <Icon
              className="closeIcon"
              onClick={[Function]}
              type="close"
            >
              <svg
                className="closeIcon"
                height="12"
                onClick={[Function]}
                viewBox="0 0 12 12"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.7803 1.28033C12.0732 0.987437 12.0732 0.512563 11.7803 0.21967C11.4874 -0.0732233 11.0126 -0.0732233 10.7197 0.21967L6 4.93934L1.28033 0.21967C0.987437 -0.0732233 0.512563 -0.0732233 0.21967 0.21967C-0.0732233 0.512563 -0.0732233 0.987437 0.21967 1.28033L4.93934 6L0.21967 10.7197C-0.0732233 11.0126 -0.0732233 11.4874 0.21967 11.7803C0.512563 12.0732 0.987437 12.0732 1.28033 11.7803L6 7.06066L10.7197 11.7803C11.0126 12.0732 11.4874 12.0732 11.7803 11.7803C12.0732 11.4874 12.0732 11.0126 11.7803 10.7197L7.06066 6L11.7803 1.28033Z"
                />
              </svg>
            </Icon>
          </div>
        </ModalHeader>
        <div
          className="content"
        >
          Any internal modal text
          <ModalActions
            cancelLabel="Cancel"
            closeModal={[Function]}
            confirmLabel="Confirm"
            onConfirmClick={[Function]}
          >
            <div
              className="actions"
            >
              <Button
                accent="negative"
                as="button"
                border={true}
                className="btnCancel"
                fluid={false}
                isActive={false}
                onClick={[Function]}
              >
                <button
                  className="btnCancel button undefined negative bordered"
                  onClick={[Function]}
                >
                  Cancel
                </button>
              </Button>
              <Button
                accent="positive"
                as="button"
                border={false}
                className=""
                fluid={false}
                isActive={false}
                onClick={[Function]}
                variant="fill"
              >
                <button
                  className=" button fill positive"
                  onClick={[Function]}
                >
                  Confirm
                </button>
              </Button>
            </div>
          </ModalActions>
        </div>
      </div>
    </Panel>
    <div
      className="dimmed"
      onClick={[Function]}
    />
  </Portal>
</Modal>
`;
